host.py

```diff
@@ -296,20 +296,43 @@ def copy_to(self, src_file: str, dst_file: str) -> None:
     def need_sudo(self) -> None:
         self.sudo_needed = True
 
-    def run(self, cmd: str, log_level: int = logging.DEBUG, env: Dict[str, str] = os.environ.copy()) -> Result:
+    def run(
+        self,
+        cmd: str,
+        log_level: int = logging.DEBUG,
+        env: Optional[Dict[str, str]] = None,
+        *,
+        env_extra: Optional[Dict[str, str]] = None,
+    ) -> Result:
         if self.sudo_needed:
             cmd = "sudo " + cmd
 
         logger.log(log_level, f"running command {cmd} on {self._hostname}")
         if self.is_localhost():
-            ret_val = self._run_local(cmd, env)
+            ret_val = self._run_local(cmd, env, env_extra=env_extra)
         else:
-            ret_val = self._run_remote(cmd, log_level)
+            if env is not None:
+                raise ValueError("remote host does not support full environment. Use env_extra instead")
+            ret_val = self._run_remote(cmd, log_level, env_extra=env_extra)
 
         logger.log(log_level, ret_val)
         return ret_val
 
-    def _run_local(self, cmd: str, env: Dict[str, str]) -> Result:
+    def _run_local(
+        self,
+        cmd: str,
+        env: Optional[Dict[str, str]],
+        *,
+        env_extra: Optional[Dict[str, str]] = None,
+    ) -> Result:
+        if env_extra:
+            if env is None:
```

Comment:
> I don't believe there is any part of the code currently that relies on this default.

Right. Because currently

- `_run_remote()` and `needs_sudo` modes don't support it anyway. So such callers would never rely on anything what `env` does.
- for `_run_local()`, leaving `env` at None is the same as keeping the current environment. That is mostly (not exactly) the same as just using the default from `os.environment.copy()`.

> Why not just set this as the default value?

What do you mean with "this"?

Reply:
Default for env can be None.

In instances where we need the local environment to be set in the run, we can set env = os.environ.copy(), and pass this to the run.

If we need additional args we can add these to env. If we only want a subset of the local env (i.e. we dont want to set PWD or PATH), we can just pass in a dict for env.

To me adding extra_args just adds complexity.

Reply:
> we can set env = os.environ.copy(), and pass this to the run.

As said, we cannot.

The code also does not and cannot handle that. How would `sudo_needed` or `_run_remote()` handle a `env` parameter?